### PR TITLE
NWM analysis tutorials

### LINKED ISSUE #66: Move NWM analysis from hytest_notebook_tutorials
Copy and refactor into `hytest-org/hytest` repo.

### PR DESCRIPTION
**Pull request by gzt5142**
closes #66 

Move and refactor NWM benchmark analysis tutorial notebooks.

### CODE REVIEW

**Comment by gzt5142** [comment]
@amsnyder 

In looking at the analysis and viz notebooks together -- the visualizations include a D-Score benchmark, but we have no notebook example for that in the repos that I've seen.  I can easily build it, using the NWM analysis notebook as a model... just changes the metrics to calc, so will be straightforward. 

Question is... do we want the viz notebook to do both of those metrics, or do a viz for each benchmark (NWM standard, d-score)?

**Comment by sfoks** [comment]
@gzt5142 

I would like the viz to do both of those metrics initially. There may be times when d-score isn't applicable but perhaps we can cross that bridge when we get there? I guess I am open to either approach. @thodson-usgs might have thoughts on this. 

additionally, Tim has a notebook on d-score with the scorecard viz, https://github.com/thodson-usgs/dscore/blob/main/notebooks/dask_demo.ipynb if that helps!

**Comment by thodson-usgs** [comment]
For now, a single viz notebook with some interactive plots and score cards is a reasonable place to start.

I think ultimately the real benchmark reports will almost always evaluate two or more models simultaneously, and we'll need to think about how best to distill multiple variables and models into a single summary (e.g., ILAMB).

I hadn't given too much thought to d-score yet. It's up to @sfoks, but I considered it lower priority than the work to setup hytest-org infrastructure and the aws bucket. If we get the critical datasets on the cloud, and really thoughtfully setup one complete analysis pipeline, I think other team members will be nicely staged to contribute more content, once the next FY kicks off.

**Comment by gzt5142** [comment]
The viz that I've got here is a collection of scatter plots, box plots, and a map, symbolized by metric. 

The score card within Tim's d-score...  would you like that score card to be in that same mega-viz notebook or does it make sense to have a viz/explore with interactivity in one notebook, and scorecards separate.   

I get that they kind of go together... I'm thinking of this in terms of tutorial/education.  There is an advantage to having more notebooks that are shorter and more focused in terms of the education content and learning objectives. However... if the skill to learn is both viz/card together, then it makes perfect sense to tutorial them that way.

**Comment by thodson-usgs** [comment]
Start by modifying https://github.com/thodson-usgs/dscore/blob/main/notebooks/dask_demo.ipynb to use the dataset on aws. I don't think we need a "mega-viz" notebook necessarily, but it would be appropriate to include a d-score scorecard together with a d-score interactive map, IMO. Scorecard gives a summary, the map lets you drill down into more detail.

**Comment by gzt5142** [comment]
@thodson-usgs 

Side question for you related to d-score..... I suspect that you may already know about this issue -- I'm trying to make sure that *I'm* doing it correctly as I port the d-score code into the same framing as this NWM analysis notebook.

**Variance** -- Not all `var()` do the same thing.  In particular, the DDOF is different for different invocations from different libraries. 
```
np.var(array_like)  ## assumes ddof=0

array_like.var()   ## assumes ddof=1, depending on what kind of object array_like is. 
```

If `array_like` is a pandas series, then DDOF is one.  The reason I bring it up is that variance computations within NSE (for example) in the NWM notebooks specifically used numpy in the original notebooks, so I was careful to explicitly specify `ddof=0` in all cases so that output values match. This revealed itself in testing, when some datasets gave a very slightly different value for NSE, which was traced to the difference in variance calculation. 

Anyway... I notice that in the D-Score code `array_like.var()` is used rather than `np.var(array_like)`.  In terms of explicit and readable code, I would like to specify the ddof in the invocation -- meaning it won't matter which lib is used... and I want to make sure I specify the one you want (0 or 1). Either of these will give you what you have now:

```
np.var(array_like, ddof=1)
# or
array_like.var(ddof=1)
```

**Comment by thodson-usgs** [comment]
Interesting, I didn't realize that about `np.var`. Explicit is good. I prefer `array_like.var(ddof=1)` as it would default to the correct behavior.

**Comment by gzt5142** [comment]
OK... getting closer, I think.  

Still need to get the scorecards sorted.  Right now, the viz notebook does the sequence of interactive plots for the NWM standard suite, then again for the dscore metrics.

**Comment by thodson-usgs** [comment]
> Related to scorecards for dscore... the example notebook has data for NWM and NHM. I'm not seeing the NHM dataset in the intake catalog to source from AWS/S3.

@sfoks @amsnyder We want NHM on aws, right?

**Comment by amsnyder** [comment]
> > Related to scorecards for dscore... the example notebook has data for NWM and NHM. I'm not seeing the NHM dataset in the intake catalog to source from AWS/S3.
> 
> @sfoks @amsnyder We want NHM on aws, right?

Yes - I need to check in with Parker about that. It might be a little delay before it gets onto S3 because I think it needs rechunked. We had a task for that but it wasn't an issue - I just turned it into an issue (#71) so I'll close #70

**Comment by gzt5142** [comment]
The ilamb-style scorecards from dscore depend on `seaborn`.  There was a question about whether we need to hold that dependency, or if the same functionality could be produced using libs we already depend on (hv, matplotlib).  I've got a couple of options built at

https://github.com/hytest-org/hytest/blob/gt-066-nwm-analysis/model_evaluation/tutorials/nwm/DScore-ScoreCard.ipynb

You will need to check out that branch and run the notebook to get the hvplot to build so you can see what it will look like in reality.

I'll keep the scorecard notebook separate until we can talk about which you prefer. Then will merge into the viz notebook. 

Scorecards are now stalled until some cloud-accessible data can be found for NHM.. which will add a column to the scorecard.

**Comment by gzt5142** [comment]
Marked ready for review -- there are a couple of tweaks to make to these notebooks once NWM data is available. In the mean while, this should be good to go for what we have available now.

**Comment by thodson-usgs** [comment]
> The ilamb-style scorecards from dscore depend on `seaborn`. There was a question about whether we need to hold that dependency, or if the same functionality could be produced using libs we already depend on (hv, matplotlib). I've got a couple of options built at

Nice @gzt5142, I'll add your matplotlib version upstream, then revise the d-score notebook

File: model_evaluation/NWM_Variables.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+metadata:
```
Comment (amsnyder) [inline model_evaluation/NWM_Variables.yml:1]: delete or move to temp folder

**Comment by gzt5142** [comment]
This PR is (I think) ready to go... the last piece was to get the data prep phase sorted out, including the upstream NWM streamflow location. 

A couple of notes to make clear : 

- Sidebars:   The `{sidebar}` directive is included in a couple of places.  This is not a markdown
  item, so it will render plain in a standard notebook.  When these get eventually published with
  jupyterbook, the sidebar will float to the right margin and become a proper sidebar. 
- I used sidebars kind of like footnotes.  For example -- if I mention rechunking, I put in a 
  sidebar with a link to the rechunking primer, rather than describing in any detail what's happening
  with the rechunk. 
- Some of those primer documents are not yet formalized -- I don't know the full filename and 
  path yet (they are part of other branches/PRs).  In the mean while, the link is to `/dev/null`
  as a placeholder. This will make these null links easy to find and correct later (search for 'null').

**Comment by gzt5142** [comment]
If new review comments come in, I'll address them as quickly as I can.  If none, I will merge at EOB today.

**Pull request merged**